### PR TITLE
Mmca 4901 : Improve the existing cash account transactions request flow - Part 3

### PR DESCRIPTION
**Pull request by HariHmrc**

### CODE REVIEW

**Comment by HariHmrc** [comment]
Welsh translation is not given in ticket, Once Sujay is back, have to get it corrected. For now, used google translate to cover up the gap.

File: app/viewmodels/RequestedTooManyTransactionsViewModel.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels
+
+import config.AppConfig
+import utils.Utils.*
```
Comment (Neozxc) [inline app/viewmodels/RequestedTooManyTransactionsViewModel.scala:20]: Please import just needed ones:

import utils.Utils.{LinkComponentValues, linkComponent}

File: app/views/cash_account_requested_too_many_transactions.scala.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+@(
+        viewModel: RequestedTooManyTransactionsViewModel
+)(
+        implicit request: Request[_],
+        messages: Messages,
+        appConfig: AppConfig
+)
+
+@main_template(pageTitle = Some(messages("cf.cash-account.detail.title")),
```
Comment (tyagikumarvineet) [inline app/views/cash_account_requested_too_many_transactions.scala.html:34]: Please retrieve title from the viewmodel as it is already present in the viewmodel

File: app/views/cash_account_requested_too_many_transactions.scala.html [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@import viewmodels.RequestedTooManyTransactionsViewModel
+
+@this(
+        main_template: Layout,
+        h1: components.h1,
```
Comment (tyagikumarvineet) [inline app/views/cash_account_requested_too_many_transactions.scala.html:21]: Remove this unused parameter and also on line 22 and 23

File: test/controllers/SelectedTransactionsControllerSpec.scala [hunk 4]
```diff
@@ -36,137 +36,227 @@ import scala.concurrent.Future
...
+
+        when(mockCustomsFinancialsApiConnector.postCashAccountStatementRequest(any, any, any, any)(any))
+          .thenReturn(Future.successful(Left(ExceededMaximum)))
+
+        val request: FakeRequest[AnyContentAsEmpty.type] =
+          fakeRequest(POST, routes.SelectedTransactionsController.onSubmit().url)
+
+        running(app) {
+          val result = route(app, request).value
+          status(result) mustBe SEE_OTHER
```
Comment (tyagikumarvineet) [inline test/controllers/SelectedTransactionsControllerSpec.scala:224]: Better to give a line break before this line to distinguish between result and assertions

File: test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  private def shouldProduceCorrectToDateInStatement(statement: String,
+                                                    toDate: LocalDate)(implicit messages: Messages): Assertion = {
+    statement.contains(s"${messages(s"month.${toDate.getMonthValue}")} ${toDate.getYear}") mustBe true
+  }
+
+  trait Setup {
+
+    val fromDate = LocalDate.parse("2020-07-18")
```
Comment (tyagikumarvineet) [inline test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala:77]: Please add return type to the public members at this and at line 78 and 79

File: test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    val toDate = LocalDate.parse("2020-07-20")
+    val selectedTxnUrl = controllers.routes.SelectedTransactionsController.onPageLoad().url
+
+    val app: Application = application.build()
+
+    implicit val appConfig: AppConfig = app.injector.instanceOf[AppConfig]
+    implicit val msgs: Messages = messages(app)
+
```
Comment (tyagikumarvineet) [inline test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala:85]: Better to remove this redundant line break

File: test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import java.time.LocalDate
+
+class RequestedTooManyTransactionsViewModelSpec extends SpecBase {
+
+  "apply method" should {
+
+    "return correct view model object" in new Setup {
+
+      val viewModel = RequestedTooManyTransactionsViewModel(fromDate, toDate, selectedTxnUrl, selectedTxnUrl)
```
Comment (tyagikumarvineet) [inline test/viewmodels/RequestedTooManyTransactionsViewModelSpec.scala:34]: Better to add a return type for this public member to make it private

File: test/views/CashAccountRequestedTooManyTransactionsSpec.scala [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  private def shouldContainCorrectTryAgainLink(implicit view: Document): Assertion = {
+    view.getElementsByClass("govuk-body govuk-!-margin-bottom-9").text() mustBe
+      messages("cf.cash-account.transactions.requested.tryAgain")
+  }
+
+  trait Setup {
+
+    val fromDate = LocalDate.parse("2020-07-18")
```
Comment (tyagikumarvineet) [inline test/views/CashAccountRequestedTooManyTransactionsSpec.scala:73]: Please add return type for the public members (at line 74, 75 and 77 as well)

File: test/views/CashAccountRequestedTooManyTransactionsSpec.scala [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    "display correct contents with heading and statement message" in new Setup {
+
+      implicit val view: Document = viewDoc
+
+      val statementMsg: String = view.getElementById("requested-too-many-transactions-message-statement").text()
+
+      titleShouldBeCorrect(view, titleMsgKey = "cf.cash-account.detail.title")
+
+      shouldContainBackLinkUrl(view, selectedTxnUrl)
```
Comment (AdamCoops) [inline test/views/CashAccountRequestedTooManyTransactionsSpec.scala:39]: I don't think you need a line break between each shouldContain

File: app/viewmodels/RequestedTooManyTransactionsViewModel.scala [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import views.html.components.{h1, p1}
+import play.api.i18n.Messages
+
+import java.time.LocalDate
+
+case class RequestedTooManyTransactionsViewModel(pageTitle: String,
+                                                 backLink: String,
+                                                 heading: HtmlFormat.Appendable,
+                                                 tryAgainLink: HtmlFormat.Appendable,
+                                                 statement01: HtmlFormat.Appendable)
```
Comment (wg-hmrc) [inline app/viewmodels/RequestedTooManyTransactionsViewModel.scala:31]: the name `statement01` feels less descriptive maybe something more meaningful

File: app/views/cash_account_requested_too_many_transactions.scala.html [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@import viewmodels.RequestedTooManyTransactionsViewModel
+
+@this(
```
Comment (tyagikumarvineet) [inline app/views/cash_account_requested_too_many_transactions.scala.html:19]: As now its only one parameter, better to write as below

@this(main_template: Layout)

File: app/viewmodels/RequestedTooManyTransactionsViewModel.scala [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                                                 heading: HtmlFormat.Appendable,
+                                                 tryAgainLink: HtmlFormat.Appendable,
+                                                 infoSection: HtmlFormat.Appendable)
+
+object RequestedTooManyTransactionsViewModel {
+
+  def apply(from: LocalDate,
+            to: LocalDate,
+            tryAgainUrl: String,
+            backUrl: String)(implicit messages: Messages, config: AppConfig): RequestedTooManyTransactionsViewModel = {
```
Comment (tyagikumarvineet) [inline app/viewmodels/RequestedTooManyTransactionsViewModel.scala:39]: Seems, config is not being used. Could you please check and remove if not used?

**Review by tyagikumarvineet — APPROVED** [review]

**Pull request merged**